api/buckets/service.go

```diff
@@ -1166,24 +1203,63 @@ func getLink(lnks []*ipld.Link, name string) *ipld.Link {
 // updateOrAddPin moves the pin at from to to.
 // If from is nil, a new pin as placed at to.
 func (s *Service) updateOrAddPin(ctx context.Context, from, to path.Path) error {
+	key := ownerFromContext(ctx)
+	a, err := s.Collections.Accounts.Get(ctx, key)
+	if err != nil {
+		return fmt.Errorf("getting account from context: %s", err)
+	}
+
+	toSize, err := s.dagSize(ctx, to)
+	if err != nil {
+		return fmt.Errorf("getting size of destination dag: %s", err)
+	}
+	if s.BucketMaxSize > 0 && toSize > s.BucketMaxSize {
+		return fmt.Errorf("bucket size is greater than max allowed size")
+	}
+
+	fromSize, err := s.dagSize(ctx, from)
+	if err != nil {
+		return fmt.Errorf("getting size of current dag: %s", err)
+	}
+
+	// Calculate if adding or moving the pin leads to a violation of
+	// total buckets size allowed in configuration.
+	futureSize := a.BucketsTotalSize - fromSize + toSize
+	if s.BucketsTotalMaxSize > 0 && futureSize > s.BucketsTotalMaxSize {
+		return ErrBucketsTotalSizeExceedsMaxSize
+	}
+
 	if from == nil {
-		if err := s.IPFSClient.Pin().Add(ctx, to); err != nil {
-			return err
-		}
-	} else {
-		if err := s.IPFSClient.Pin().Update(ctx, from, to); err != nil {
-			if err.Error() == pinNotRecursiveMsg {
-				if err = s.IPFSClient.Pin().Add(ctx, to); err != nil {
-					return err
-				}
-			} else {
-				return err
-			}
+		return s.IPFSClient.Pin().Add(ctx, to)
+	}
+
+	if err := s.IPFSClient.Pin().Update(ctx, from, to); err != nil {
+		if err.Error() == pinNotRecursiveMsg {
+			return s.IPFSClient.Pin().Add(ctx, to)
 		}
+		return err
 	}
+
+	if err := s.Collections.Accounts.SetBucketsTotalSize(ctx, key, futureSize); err != nil {
+		return fmt.Errorf("updating new buckets total size: %s", err)
+	}
+
 	return nil
 }
 
+// dagSize returns the cummulative size of root. If root is nil, it returns 0.
+func (s *Service) dagSize(ctx context.Context, root path.Path) (int64, error) {
+	if root == nil {
+		return 0, nil
+	}
```

Comment:
Clarified the `root==nil` case in comments. This was to make smoother the `updateOrAddPin` if the `to` was `nil`.